### PR TITLE
Projections

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add projection support for MergeTree* tables. 

Detailed description / Documentation draft:

The documentation is almost listed in  https://github.com/ClickHouse/ClickHouse/issues/14730

Possible TODOs.

- [x] Complete query analysis for projections.
- [x] Normal projections with different order by clause
- [ ] ~~ProjectionMergeTree to store projection data only~~ can be done via CollapsingMergeTree with default sign = 1
- [ ] Secondary indices via projection

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> Some exception occurred :( 

`GithubException: 500 null`
This is GitHub downtime, https://www.githubstatus.com/history

Need to restart the tests.

**Comment by zhaojintaozhao** [comment]
> I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en
> 
> Changelog category (leave one):
> 
> * New Feature
> 
> Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> 
> Add projection support for MergeTree* tables.
> 
> Detailed description / Documentation draft:
> 
> The documentation is almost listed in #14730
> 
> Possible TODOs.
> 
> * [ ]  Complete query routing for projections. Currently alias in group-by will not work.
> * [ ]  Normal projections with different order by clause
> * [ ]  ProjectionMergeTree to store projection data only
> * [ ]  Secondary indices via projection

@amosbird : 
This is a very valuable feature;
I have some questions about projects.
1. What is the difference between projections and materialized views?
2. Does projections also use the idea of kylin cube? Pre-aggregate the query results in advance. The aggregated results are directly queried during query. Projections: Are the combinations of all dimensions calculated in advance?

**Comment by itzikiusa** [comment]
Long expected feature!
Waiting for PR to be approved, and the following steps, specially different order by
Thank you very much for your work!

**Comment by amosbird** [comment]
```
2021-05-05 23:51:05 00992_system_parts_race_condition_zookeeper_long:                       Received exception from server (version 21.6.1):
2021-05-05 23:51:05 Code: 999. DB::Exception: Received from localhost:9000. DB::Exception: Run time inconsistency, path: /clickhouse/task_queue/ddl/query-0000003330. 
2021-05-05 23:51:06 Received exception from server (version 21.6.1):
2021-05-05 23:51:06 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: There was an error on [localhost:19000]: Code: 81, e.displayText() = DB::Exception: Database test_06rrwt doesn't exist (version 21.6.1.6765). 
2021-05-05 23:51:06 [ FAIL ] 4.53 sec. - return code 81
2021-05-05 23:51:06 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: Database `default` doesn't exist. 
2021-05-05 23:51:06 
2021-05-05 23:51:06 , result:
2021-05-05 23:51:06 
2021-05-05 23:51:06 
```

**Comment by amosbird** [comment]
> Integration tests (asan) — Timeout, fail: 0, passed: 1314, flaky: 1
What does this mean?

**Comment by KochetovNicolai** [comment]
Currently, most of mentioned in https://github.com/ClickHouse/ClickHouse/issues/14730#issue-698305277 is implemented. Exceptions are: `WHERE` in projection definition and everything from `Projection Analysis and Settings`.

This feature is experimental. Enable `allow_experimental_projection_optimization` to use projections in `SELECT`.
Projection with `ORDER BY` uses it's own primary key when selected. Read-in-order optimization also supported. 
Projection with `GROUP BY` uses stored pre-aggregated data.

Alter update/delete can't be used together with projections (currently).

There is still not enough tests. Also, not all the code was thoughtfully reviewed. However, I am going to merge this pr, because the feature works in general (and to avoid major conflicts).

**Comment by amosbird** [comment]
https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/integration_tests_(release)/integration_run_test_cluster_copier_test_py_1.txt

cluster copier test throws code 85: `FORMAT_IS_NOT_SUITABLE_FOR_INPUT`

**Comment by amosbird** [comment]
https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/integration_tests_(thread)/integration_run_test_jbod_balancer_test_py_0.txt

insert timeout

File: tests/queries/skip_list.json [hunk 1]
```diff
@@ -392,6 +392,9 @@
         "01475_read_subcolumns_storages",
         "01674_clickhouse_client_query_param_cte",
         "01666_merge_tree_max_query_limit",
+        "01710_projections",
```
Comment (abyss7) [inline tests/queries/skip_list.json:395]: Why don't you implement parsing for ANTLR grammar?

Comment (amosbird) [reply tests/queries/skip_list.json:395]: Done

**Comment by alexey-milovidov** [comment]
@KochetovNicolai There is no comment about `MergeTreeDataSelectCache`.

You've missed the case when PVS-Studio argued about uncomprehensible code:
> PVS check — Have 7 new errors, 21 total errors 

https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/pvs_studio_report/pvs-studio-html-report/sources/MergeTreeDataSelectExecutor.cpp_13.html#ln583

The code is correct but it becomes too tangled, which PVS-Studio does not like.

**Comment by alexey-milovidov** [comment]
We need clear explanation - what MergeTreeDataSelectCache is doing, what's going on in the code...

**Comment by olgarev** [comment]
Internal ticket DOCSUP-12894

**Comment by zhouerlin** [comment]
Currently, projection does not support the where clause,
What is the difficulty in implementing it or for what purpose it was designed to do so?

**Pull request merged**